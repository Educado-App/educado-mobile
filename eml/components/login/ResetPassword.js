import { View } from 'react-native';
import React, { useState, useEffect } from 'react';
import FormTextField from '../general/forms/FormTextField';
import FormButton from '../general/forms/FormButton';
import EducadoModal from '../general/EducadoModal';
import EnterNewPasswordScreen from './EnterNewPasswordScreen';
import Text from '../general/Text';
import { sendResetPasswordEmail, validateResetPasswordCode } from '../../api/userApi';
import FormFieldAlert from '../general/forms/FormFieldAlert';
import { validateEmail } from '../general/Validation';
import ToastNotification from '../general/ToastNotification';
import ShowAlert from '../general/ShowAlert';
import PropTypes from 'prop-types';

/**
 * Component to create modal (popup) that prompts user for
 * email and code from email to reset password
 * @param {Object} props Should contain the following properties
 * - modalVisible: Boolean to show if modal should be visible
 * - onModalClose: Function to do when modal closes
 */
export default function ResetPassword(props) {

	const emailAlertMessage = 'Não existe nenhum usuário com este email!';
	const [email, setEmail] = useState('');
	const [token, setToken] = useState('');
	const [emailSent, setEmailSent] = useState(false);
	const [codeEntered, setCodeEntered] = useState(false);
	const [passwordResetAlert, setPasswordResetAlert] = useState('');
	const [tokenAlert, setTokenAlert] = useState('');
	const [isSuccess, setIsSuccess] = useState(false);
	const [buttonLoading, setButtonLoading] = useState(false);

	/**
   * Function to display an alert to the user
   * @param {String} message Message to display in alert
   * @param {Boolean} status Displays if alert is success or error (green if true or red if false)
   */
	function displayErrorAlert(message, status) {
		setPasswordResetAlert(message);
		setIsSuccess(status);
	}

	useEffect(() => {
		if (email === '') {
			displayErrorAlert('', false);
			return;
		}

		const validationError = validateEmail(email);
		displayErrorAlert(validationError, false);
	}, [email]);

	/**
   * Function to send mail to user with code to reset password
   * @param {*} email 
   */
	async function sendEmail(email) {
		const obj = {
			email,
		};

		setButtonLoading(true);
		await sendResetPasswordEmail(obj)
			.then(async () => {
				setEmailSent(true);
				ToastNotification('success', 'E-mail enviado!'); //email sent!
			}).catch((error) => {
				switch (error?.error?.code) {
				case 'E0401':
					// No user exists with this email!
					displayErrorAlert(emailAlertMessage, false);
					break;

				case 'E0406':
					// Too many resend attempts!
					displayErrorAlert('Muitas tentativas de reenvio! Espere 5 minutos...', false);
					break;

				case 'E0004':
					// User not found!
					displayErrorAlert('Usuário não encontrado!', false);
					break;

					// TODO: What error should we give here instead? Unknown error? 
				default:
					// Errors not currently handled with specific alerts
					displayErrorAlert('Erro desconhecido!', false);
				}
			});
		setButtonLoading(false);
	}


	/**
   * Function to validate the code entered by the user
   * @param {String} email 
   * @param {String} token 
   */

	async function validateCode(email, token) {
		const obj = {
			email,
			token,
		};

		await validateResetPasswordCode(obj)
			.then(async () => {
				setCodeEntered(true);
			}).catch((error) => {
				switch (error?.error?.code) {
				case 'E0401':
					// No user exists with this email!
					displayErrorAlert(emailAlertMessage, false);
					break;

				case 'E0404':
					// Code expired!
					setTokenAlert('Código expirado!');
					break;

				case 'E0405':
					// Incorrect code!
					setTokenAlert('Código incorreto!');
					break;

				default:
					// Errors not currently handled with specific alerts
					ShowAlert('Erro desconhecido!');
					console.log(error);
				}
			});
	}

	//resets the state of the reset password modal
	const resetState = () => {
		setEmailSent(false);
		setCodeEntered(false);
		displayErrorAlert('', false);
		setTokenAlert('');
		setToken('');
	};

	//checks if the 4-digit code entered is valid
	const codeInputValid = (code) => {
		return /^\d+$/.test(code) && (code.length === 4);
	};

	return (
		<EducadoModal modalVisible={props.modalVisible} closeModal={props.onModalClose} id="EducadoModal" title="Redefinição de senha">
			<View className="my-[80px] px-10">
				{!codeEntered ? (
					<View>
						<FormTextField
							bordered={true}
							placeholder="Insira sua e-mail"
							label="E-mail"
							required={true}
							onChangeText={(email) => setEmail(email)}
							keyboardType="email-address"
							testId="emailInput"
							value={email}
						/>
						<FormFieldAlert testId="emailAlert" label={passwordResetAlert} success={isSuccess} />
						<View className="mt-[40px]">
							{emailSent ? (
								<View>
									<Text className="text-left mb-[10px]">
										{/* We have sent a code to your mail to reset your password,
                     please enter the code you have received below: */}
                    Enviamos para o seu email um código de redefinição de senha. Insira o código abaixo.
									</Text>
									<FormTextField
										bordered={true}
										placeholder="X X X X"
										onChangeText={(token) => setToken(token)}
										testId="tokenInput"
										value={token}
									/>
									<FormFieldAlert testId="tokenAlert" label={tokenAlert} />
									{/* Continue button */}
									<View className="mt-[40px] mb-[24px]">
										<FormButton
											// Continue 
											label={buttonLoading ? 'Validando código...' : 'Continuar'}
											onPress={() => validateCode(email, token)}
											testId="validateCodeBtn"
											disabled={!codeInputValid(token)}
										>
											{buttonLoading ? 'Validando código...' : 'Continuar'}
										</FormButton>
									</View>
									<View className="mx-10 flex-row justify-center">
										{/* Did not receieve the code? */}
										<Text>O código não chegou?</Text>
										{/* Resend code*/}
										<Text className="underline ml-1" onPress={() => sendEmail(email)}>Reenviar código</Text>
									</View>
								</View>
							) : (
								<FormButton
									// Send code
									onPress={() => sendEmail(email)}
									testId="resetPasswordButton"
									disabled={passwordResetAlert !== '' || email === '' || buttonLoading}
								>
									{buttonLoading ? 'Enviando e-mail...' : 'Enviar código'}
								</FormButton>
							)}
						</View>
					</View>
				) : (
					<EnterNewPasswordScreen
						email={email}
						token={token}
						hideModal={props.onModalClose}
						resetState={resetState}
					/>
				)}
			</View>
		</EducadoModal>
	);
}

ResetPassword.propTypes = {
	modalVisible: PropTypes.bool,
	onModalClose: PropTypes.func,
};