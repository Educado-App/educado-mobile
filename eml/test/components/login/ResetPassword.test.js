import renderer from 'react-test-renderer';
import React from 'react';
import ResetPassword from '../../../components/login/ResetPassword';

jest.mock('../../../api/userApi', () => ({
  sendResetPasswordEmail: jest.fn(async ({ email }) => {
    if (email !== 'test@test.dk') {
      return Promise.reject({ error: { code: 'E0401' } }); // No user exists with this email!
    } else if (email !== 'resend@test.dk') {
      return Promise.reject({ error: { code: 'E0406' } }); // Too many resend attempts!
    } else {
      return Promise.resolve({ data: 'Email sent' });
    }
  }),
  validateResetPasswordCode: jest.fn(async ({ email, token }) => {
    if (email !== 'test@test.dk' && email !== 'expired@test.dk') {
      return Promise.reject({ error: { code: 'E0401' } }); // No user exists with this email!
    } else if (email === 'test@test.dk' && token !== '1234') {
      return Promise.reject({ error: { code: 'E0405' } }); // Invalid token!
    } else if (email === 'expired@test.dk') {
      return Promise.reject({ error: { code: 'E0404' } }); // Code expired!
    } else {
      return Promise.resolve({ data: 'Code validated' });
    }
  }),
}));

describe('ResetPassword', () => {

  describe('Test modal open and close', () => {

    let resetPassword;

    it('Ensure that the reset password page is rendered correctly when not visible', async () => {
      await renderer.act(() => {
        resetPassword = renderer.create(<ResetPassword modalVisible={false} onModalClose={() => { }} />);
      });
      let tree = resetPassword.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('Ensure that the reset password page is rendered correctly when visible', async () => {
      await renderer.act(() => {
        resetPassword = renderer.create(<ResetPassword modalVisible={true} onModalClose={() => { }} />);
      });
      let tree = resetPassword.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Test errors and validation', () => {

    let resetPassword;
    let emailInput;
    let resetPasswordButton;
    let emailAlert;
    let tokenInput;
    let validateCodeBtn;
    let tokenAlert;

    // Function used to initialize tokenInput and validateCodeBtn
    // this should be called after rendering new page, by pressing resetPasswordButton
    function initTokenComponents() {
      tokenInput = resetPassword.root.findByProps({ testId: 'tokenInput' });
      validateCodeBtn = resetPassword.root.findByProps({ testId: 'validateCodeBtn' });
      tokenAlert = resetPassword.root.findByProps({ testId: 'tokenAlert' });
    }

    beforeAll(async () => {
      await renderer.act(async () => {
        resetPassword = renderer.create(<ResetPassword />);
      }).then(() => {
        emailInput = resetPassword.root.findByProps({ testId: 'emailInput' });
        resetPasswordButton = resetPassword.root.findByProps({ testId: 'resetPasswordButton' });
        emailAlert = resetPassword.root.findByProps({ testId: 'emailAlert' });
      });
    });

    it('Error if email does not exist', async () => {
      // Ensure emailAlert is properly initialized
    
      await renderer.act(async () => {
        await emailInput.props.onChangeText('test@test.com');
        await resetPasswordButton.props.onPress();
      }).then(() => {
        // Initialize emailAlert here after the component has rendered
        emailAlert = resetPassword.root.findByProps({ testId: 'emailAlert' });
    
        expect(emailAlert.props.label).not.toBe('');
      });
    });
    

    it('Error if invalid email, no error if valid', async() => {
      emailInput = resetPassword.root.findByProps({ testId: 'emailInput' });
      await renderer.act(() => {
        emailInput.props.onChangeText('test.com');
      });
      expect(emailAlert.props.label).not.toBe('');

      await renderer.act(() => {
        emailInput.props.onChangeText('test@test.c');
      });
      expect(emailAlert.props.label).not.toBe('');

      await renderer.act(() => {
        emailInput.props.onChangeText('test@test');
      });
      expect(emailAlert.props.label).not.toBe('');

      await renderer.act(() => {
        emailInput.props.onChangeText('test@test.com');
      });
      expect(emailAlert.props.label).toBe('');
    });


    it('Error message with too many attempts', async () => {
      await renderer.act(async () => {
        await emailInput.props.onChangeText('resend@test.dk');
        await resetPasswordButton.props.onPress();
      }).then(() => {
        expect(emailAlert.props.label).not.toBe('');
      });
    });

    it('Error message with invalid token', async () => {
      await renderer.act(async () => {
        await emailInput.props.onChangeText('test@test.dk');
        await resetPasswordButton.props.onPress();

        initTokenComponents();
        await tokenInput.props.onChangeText('1337');
        await validateCodeBtn.props.onPress();
      }).then(() => {
        expect(tokenAlert.props.label).not.toBe('');
      });
    });

    it('Error message with expired token', async () => {
      await renderer.act(async () => {
        await emailInput.props.onChangeText('expired@test.dk');
        await resetPasswordButton.props.onPress();

        // initialize after rendering new page, by pressing resetPasswordButton
        initTokenComponents();
        await tokenInput.props.onChangeText('1337');
        await validateCodeBtn.props.onPress();
      }).then(() => {
        expect(tokenAlert.props.label).not.toBe('');
      });
    });
  });
});
