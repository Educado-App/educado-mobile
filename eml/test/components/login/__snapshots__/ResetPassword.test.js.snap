// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPassword Test modal open and close Ensure that the reset password page is rendered correctly when not visible 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  style={
    [
      {
        "borderBottomWidth": 8,
        "borderLeftWidth": 8,
        "borderRightWidth": 8,
        "borderTopWidth": 8,
      },
    ]
  }
  visible={false}
/>
`;

exports[`ResetPassword Test modal open and close Ensure that the reset password page is rendered correctly when visible 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  style={
    [
      {
        "borderBottomWidth": 8,
        "borderLeftWidth": 8,
        "borderRightWidth": 8,
        "borderTopWidth": 8,
      },
    ]
  }
  visible={true}
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      style={
        [
          {
            "display": "flex",
          },
          {
            "justifyContent": "center",
          },
          {
            "paddingTop": "10%",
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "display": "flex",
            },
            {
              "flexDirection": "row",
            },
            {
              "justifyContent": "flex-end",
            },
            {
              "paddingLeft": "10%",
              "paddingRight": "10%",
            },
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": undefined,
                  "fontSize": 24,
                },
                undefined,
                {
                  "fontFamily": "entypo",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
      <View
        style={
          [
            {
              "display": "flex",
            },
            {
              "flexDirection": "row",
            },
            {
              "justifyContent": "center",
            },
            {
              "marginBottom": "10%",
              "marginTop": "10%",
            },
          ]
        }
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            [
              {
                "height": 48,
              },
            ]
          }
        />
      </View>
      <View
        style={
          [
            {
              "display": "flex",
            },
            {
              "flexDirection": "row",
            },
            {
              "justifyContent": "flex-start",
            },
            {
              "paddingLeft": "10%",
              "paddingRight": "10%",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "Montserrat-Regular",
              },
              {
                "color": "#383838",
              },
              {
                "fontSize": 16,
              },
              [
                {
                  "textAlign": "center",
                },
                {
                  "fontSize": 24,
                },
              ],
            ]
          }
        >
          Redefinição de senha
        </Text>
      </View>
    </View>
    <View
      style={
        [
          {
            "marginBottom": 80,
            "marginTop": 80,
          },
          {
            "paddingLeft": 40,
            "paddingRight": 40,
          },
        ]
      }
    >
      <View>
        <View>
          <View
            style={
              [
                {
                  "display": "flex",
                },
                {
                  "flexDirection": "row",
                },
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "fontFamily": "Montserrat-Regular",
                  },
                  {
                    "color": "#383838",
                  },
                  {
                    "fontSize": 16,
                  },
                  [
                    {
                      "marginLeft": 8,
                    },
                    {
                      "fontSize": 12,
                      "lineHeight": 16,
                    },
                  ],
                ]
              }
            >
              E-mail
            </Text>
            <Text
              style={
                [
                  {
                    "fontFamily": "Montserrat-Regular",
                  },
                  {
                    "color": "#383838",
                  },
                  {
                    "fontSize": 16,
                  },
                  [
                    {
                      "marginLeft": 4,
                    },
                    {
                      "fontSize": 12,
                      "lineHeight": 16,
                    },
                    {
                      "color": "#FF4949",
                    },
                  ],
                ]
              }
            >
              *
            </Text>
          </View>
          <View>
            <TextInput
              autoComplete="off"
              keyboardType="email-address"
              onChangeText={[Function]}
              passwordGuidelines={false}
              placeholder="Insira sua e-mail"
              secureTextEntry={false}
              style={
                [
                  {
                    "paddingBottom": 4,
                    "paddingTop": 4,
                  },
                  {
                    "paddingLeft": 10,
                  },
                  {
                    "backgroundColor": "#FFFFFF",
                  },
                  {
                    "borderBottomLeftRadius": 8,
                    "borderBottomRightRadius": 8,
                    "borderTopLeftRadius": 8,
                    "borderTopRightRadius": 8,
                  },
                  {
                    "borderBottomWidth": 2,
                    "borderLeftWidth": 2,
                    "borderRightWidth": 2,
                    "borderTopWidth": 2,
                  },
                  {
                    "borderBottomColor": "#A1ACB2",
                    "borderLeftColor": "#A1ACB2",
                    "borderRightColor": "#A1ACB2",
                    "borderTopColor": "#A1ACB2",
                  },
                ]
              }
              value=""
            />
          </View>
        </View>
        <View
          style={
            [
              {
                "flexDirection": "row",
              },
              {
                "alignItems": "center",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "fontFamily": "Montserrat-Regular",
                },
                {
                  "color": "#383838",
                },
                {
                  "fontSize": 16,
                },
                [
                  {
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                  {
                    "color": "#FF4949",
                  },
                  {
                    "marginLeft": 8,
                    "marginRight": 8,
                  },
                ],
              ]
            }
          />
        </View>
        <View
          style={
            [
              {
                "marginTop": 40,
              },
            ]
          }
        >
          <View>
            <View
              accessibilityState={
                {
                  "disabled": true,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "backgroundColor": "#5ECCE9",
                  "borderBottomLeftRadius": 8,
                  "borderBottomRightRadius": 8,
                  "borderTopLeftRadius": 8,
                  "borderTopRightRadius": 8,
                  "opacity": 0.5,
                  "paddingBottom": 16,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 16,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Montserrat-Regular",
                    },
                    {
                      "color": "#383838",
                    },
                    {
                      "fontSize": 16,
                    },
                    [
                      {
                        "textAlign": "center",
                      },
                      {
                        "fontFamily": "Montserrat-Bold",
                      },
                      {
                        "fontSize": 16,
                      },
                      {
                        "color": "#FFFFFF",
                      },
                    ],
                  ]
                }
              >
                Enviar código
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
